### PR TITLE
Refactoring and speeding-up of the XML parser for the L1T O2O

### PR DESCRIPTION
**Pull request by kkotov**
The original XML parser developed by Thomas and Vangelis did a good job for the rapidly developing L1T O2O in Run II 2016. Now I've revisited some of it's design, simplified the interface, and as the result took advantage of the c++11 move semantics. In a nutshell, I replace the std::vector&lt;l1t::TableRow&gt; table-structure with a simpler std::map of column_name -&gt; vector_of_values. While l1t::TableRow and l1t::Setting both lack a move constructors and abuse copy semantics every time a new l1t::Setting is added to the l1t::TrigSystem::procSettings_ array, the new table representation burns no extra cycles when r-value l1t::Parameter is added to the l1t::TriggerSystem::procParameters array. This change combined together with further optimization in l1t::TriggerSystem class provides a substantial (up to x5) speed increase, easily noticeable when parsing uGTrs configuration.

The original code still lives side by side with the new one in L1Trigger/L1TCommon. The O2O online producers in L1TriggerConfig/L1TConfigProducers as well as uGMT and BMTF helper classes in L1Trigger/{L1TMuon,L1TMuonBarrel} were modified to use the new parser. The code was tested by re-running L1TO2O on one of the 2016 keys and observing agreement in payload hash codes

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

@kkotov You need to make a pull request in cms-bot to add the new package.

**Comment by kkotov** [comment]
@lgray , could you point me to any instructions on how to add new packages using cms-bot? Thanks!

**Comment by lgray** [comment]
@kkotov You need to clone https://github.com/cms-sw/cms-bot/ and then edit https://github.com/cms-sw/cms-bot/blob/master/categories.py and make a pull request.

**Comment by kkotov** [comment]
@lgray  Thanks for the hint! I've just made a new PR: https://github.com/cms-sw/cms-bot/pull/799 . A theoretical question aside: am I also allowed to introduce new categories? For example, I've added my new package L1TriggerConfig/XmlConfigTools under l1 meaning that l1 managers (Mike or Vladimir) will have to sign off every time I modify my package. On the other hand, if I introduce my own l1o2o category with the packages that are not used anywhere but in L1T O2O, I'd be a natural person to have full control over this code.

**Comment by lgray** [comment]
I'd bring that up with other people, perhaps @slava77 knows? Otherwise my
best guess is to put it under L1.

**Comment by slava77** [comment]
On 1/5/17 9:47 AM, Lindsey Gray wrote:
> I'd bring that up with other people, perhaps @slava77 knows? Otherwise my
> best guess is to put it under L1.

bring this up in the ORP and then in the O&C weekly meeting or a
management meeting.

categories are not cast in stone. A new category should not be hard to
implement.
(more changes will be necessary than just in the categories.py file)
You can even make the current l1 managers also to be in the l1o2o
category so that the signature ability is available from more people.


>

**Comment by rekovic** [comment]
as discussed with @kkotov, needs moving code around and putting new modules in the existing L1Trigger/L1TCommon package.

**Comment by davidlange6** [comment]
As for categories - of course the idea is for actual code review before it gets into the release for general use - so having developers "review" their own code makes little sense..

**Comment by ggovi** [comment]
please test

**Comment by kkotov** [comment]
As agreed with @rekovic , the new code is placed back into L1TCommon side by side with the old files. While reviewing the code the two additional files were identified not updated to this branch:  L1Trigger/L1TMuon/plugins/BuildFile.xml and L1TriggerConfig/L1TConfigProducers/src/L1TMuonEndcapParamsOnlineProd.cc. Now a new test needs to be triggered.

**Comment by kkotov** [comment]
I also rebased this PR to the current branch CMSSW_9_0_X

**Comment by slava77** [comment]
-1
rebase or similar fix is needed 

https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideHATS2016Git#CMSSW_rebasing_recipe

**Comment by rekovic** [comment]
please test

**Comment by kkotov** [comment]
Judging from the logs, I don't see anything in my code that could have cause the problem with tests above. @rekovic do you think we may want to trigger another test?

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by kkotov** [comment]
@davidlange6 is there anything left for this PR before it is merged?

File: L1Trigger/L1TCommon/src/XmlConfigParser.cc [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "xercesc/util/PlatformUtils.hpp"
+
+XERCES_CPP_NAMESPACE_USE
+
+using namespace l1t;
+
+inline std::string _toString(XMLCh const* toTranscode) {
```
Comment (ianna) [inline L1Trigger/L1TCommon/src/XmlConfigParser.cc]: @kkotov - please use Xerces string utilities from http://cmslxr.fnal.gov/source/Utilities/Xerces/interface/XercesStrUtils.h

**Review by ianna — CHANGES_REQUESTED** [review]
Xerces string utilities guarantee proper memory management. Please, consider using them.

File: L1Trigger/L1TCommon/interface/Parameter.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    std::map<std::string,unsigned int> columnNameToIndex; // remember original positions of the columns
+
+public:
+    std::string getId        (void) const noexcept { return id;             }
+    std::string getProcOrRole(void) const noexcept { return procOrRole;     }
+    std::string getType      (void) const noexcept { return type;           }
+    std::string getValueAsStr(void) const noexcept { return scalarOrVector; }
+
+    bool isScalar(void) const noexcept {
+        if( type.find("vector") != std::string::npos ||
```
Comment (davidlange6) [inline L1Trigger/L1TCommon/interface/Parameter.h]: hi @kkotov - xml is always long to review - here are some comments

Comment (davidlange6) [inline L1Trigger/L1TCommon/interface/Parameter.h]: please make the "vector" and "table" into const expr.

File: L1Trigger/L1TCommon/src/Parameter.cc [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+namespace l1t{
+
+Parameter::Parameter(const char *id,
+                     const char *procOrRole,
+                     const char *type,
+                     const char *value,
+                     const char *delimeter
+                    ){
+    this->id = id;
```
Comment (davidlange6) [inline L1Trigger/L1TCommon/src/Parameter.cc]: would be more nicely handled by following a naming convention. Eg, a usual one in CMS is member data ends in '_'

File: L1Trigger/L1TCommon/src/XmlConfigParser.cc [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+
+void XmlConfigParser::readRootElement(TriggerSystem& aTriggerSystem, const std::string& sysId)
+{
+  DOMElement* rootElement = doc_->getDocumentElement();
+  if (rootElement) {
+    if (rootElement->getNodeType() == DOMNode::ELEMENT_NODE) {
+      readElement(rootElement, aTriggerSystem, sysId);
+    }
+  } else {
+    std::cout << "No xml root element found" << std::endl;
```
Comment (davidlange6) [inline L1Trigger/L1TCommon/src/XmlConfigParser.cc]: Please use the message logger or other thread safe construct instead of cout.

**Comment by davidlange6** [comment]
and generically - it looks like this PR is intended to obsolete some set of classes, hopefully including one of the two existing L1 xml readers. Is that the case? if so what is the plan to actually delete the obsolete code?

File: L1Trigger/L1TCommon/src/XmlConfigParser.cc (hunk 4 above)
Comment (kkotov) [inline L1Trigger/L1TCommon/src/XmlConfigParser.cc]: That'll bring a dependency on the Logging library of CMSSW. I don't know if Thomas Reis (the author of XmlConfigReader, that this class is derived from) had ever intended to use it in a standalone mode. I'll better try to clean all the std::cout's instead.

File: L1Trigger/L1TCommon/src/XmlConfigParser.cc (hunk 1 above)
Comment (kkotov) [inline L1Trigger/L1TCommon/src/XmlConfigParser.cc]: Thanks @ianna for pointing to this solution.

File: L1Trigger/L1TCommon/interface/Parameter.h (hunk 2 above)
Comment (kkotov) [inline L1Trigger/L1TCommon/interface/Parameter.h]: @davidlange6 I am not sure I understand what you mean here. Non of the l1t::Parameter objects are constructed or manipulated at compile-time. They are constructed as the xml configuration parsed in runtime. So declaring any of the trivial methods of this class as constexpr won't have any effect in the end.

**Comment by kkotov** [comment]
@davidlange6 , Yes, the XmlConfigParser is depricating XmlConfigReader as well as l1t::Parameters is replacing l1t::Setting and l1t::TableRow classes. I would prefer to profit (not for a long time) of having the new and old systems side by side in the release.

File: L1Trigger/L1TCommon/src/Parameter.cc (hunk 3 above)
Comment (kkotov) [inline L1Trigger/L1TCommon/src/Parameter.cc]: Thanks for pointing to the naming convention. I'd stick with the 'this->' construct for the moment so as to minimize the changes.

**Comment by rekovic** [comment]
please test

**Comment by kkotov** [comment]
The error above must be due to a problem with RelVal accessibility. But the logs seem to be incomplete.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**